src/storage/src/compacter.rs

```diff
@@ -0,0 +1,691 @@
+
+        for segment in finished_segments {
+            let batch = Batch::create(&segment, &self.trace_path)?;
+            self.batches.push(batch);
+            fs::remove_file(&segment).with_context(|| {
+                format!(
+                    "failed to remove consumed wal segment {}",
+                    segment.display()
+                )
+            })?;
+        }
+
+        Ok(())
+    }
+
+    fn find_finished_wal_segments(&self) -> Result<Vec<PathBuf>, anyhow::Error> {
+        lazy_static! {
+            static ref FINISHED_WAL_SEGMENT_REGEX: Regex =
+                Regex::new("^log-[0-9]+-final$").unwrap();
+        }
+
+        let mut segments = read_dir_regex(&self.wal_path, &FINISHED_WAL_SEGMENT_REGEX)?;
+        segments.sort_by_key(|segment| {
+            segment
+                .to_str()
+                .unwrap()
+                .split('-')
+                .nth(1)
+                .unwrap()
+                .parse::<usize>()
+                .unwrap()
+        });
+
+        Ok(segments)
+    }
+
+    fn find_unfinished_wal_segment(&self) -> Result<PathBuf, anyhow::Error> {
+        lazy_static! {
+            static ref UNFINISHED_WAL_SEGMENT_REGEX: Regex = Regex::new("^log-[0-9]+$").unwrap();
+        }
+
+        let mut segments = read_dir_regex(&self.wal_path, &UNFINISHED_WAL_SEGMENT_REGEX)?;
+        if segments.len() > 1 {
+            bail!(
+                "Expected only a single unfinished wal segment at {}. Found {}",
+                self.wal_path.display(),
+                segments.len()
+            );
+        }
+
+        if segments.len() == 0 {
+            bail!(
+                "Expected at least a single unfinished wal segment at {}. Found none.",
+                self.wal_path.display()
+            );
+        }
+
+        Ok(segments.pop().unwrap())
+    }
+
+    fn find_batches(&self) -> Result<Vec<Batch>, anyhow::Error> {
+        lazy_static! {
+            static ref BATCH_REGEX: Regex = Regex::new("^batch-[0-9]+-[0-9]+$").unwrap();
+        }
+
+        let batches = read_dir_regex(&self.trace_path, &BATCH_REGEX)?;
+
+        let mut batches: Vec<Batch> = batches
+            .into_iter()
+            .map(Batch::reinit)
+            .collect::<Result<_, _>>()
+            .unwrap();
+        batches.sort_by_key(|batch| batch.lower);
+        Ok(batches)
+    }
+
+    /// Try to compact all of the batches we know about into a single batch from
+    /// [compaction_frontier, upper)
+    ///
+    /// TODO: the approach to compacting is likely very suboptimal.
+    fn compact(&mut self) -> Result<(), anyhow::Error> {
+        self.consume_wal()?;
+
+        if self.batches.len() > 10 {
+            let batches = std::mem::replace(&mut self.batches, vec![]);
+            let batch = Batch::compact(&batches, &self.trace_path, self.compaction)?;
+            self.batches.push(batch);
+
+            // TODO: This seems like potentially a place with a weird failure mode.
```

Comment:
Any more thoughts? It seems like if the above ensures that the corresponding batches are persisted, then possibly fine? You are potentially removing information if the compaction is further along than for the batches. Maybe a great point to .. maybe write down what wants to stay true about the persisted data. I'm not sure exactly what that is.

This seems like a fine place to decrement the refcount for these batches, though others might want to keep them alive for other reasons. Could do with a comment to that effect (I think the `TODO` is roughly that: "look here for the reason data loss happened").

Reply:
One "failure mode" thing that you might guard against in the loading of traces is: you might want to read in all batch descriptions, discover the lower and upper bounds, and then *not* read batches that are subsumed by other batches. If the above fails to atomically delete all batches, then the "just read them all in" implementation fails, even though the descriptions were sufficient to avoid that problem.

Reply:
The strategy for DD batch reading (in the persistence prototype) is a dynamic program, which wants to find contiguous `[lower, upper)` intervals starting at zero, minimizing the number of batches. If you end up with something other than a dynamic program I'd be 1. interested and 2. suspicious. :D

Reply:
yeah we're absolutely going to want the same thing here -- I ran out of steam before writing that logic. I _think_ we have enough information about how compaction happens in this case to be fine with a greedy algorithm (like we know that we're not going to see arbitrarily bad intervals, because we are the ones who compacted consecutive batches together) but in general yeah

Reply:
i wrote this logic